mirage/client/dns_client_mirage.ml

```diff
@@ -5,11 +5,53 @@ module Log = (val Logs.src_log src : Logs.LOG)
 
 module IM = Map.Make(Int)
 
-module Make (R : Mirage_random.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (S : Tcpip.Stack.V4V6) = struct
+module type S = sig
+  module Transport : Dns_client.S
+    with type io_addr = [ `Plaintext of Ipaddr.t * int | `Tls of Tls.Config.client * Ipaddr.t * int ]
+     and type +'a io = 'a Lwt.t
+
+  include module type of Dns_client.Make(Transport)
+
+  val nameserver_of_string : string -> (Transport.io_addr, [> `Msg of string ]) result
 
+  val connect :
+    ?nameservers:string list ->
+    ?timeout:int64 ->
+    Transport.stack -> t Lwt.t
+end
+
+module Make (R : Mirage_random.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (S : Tcpip.Stack.V4V6) = struct
   module TLS = Tls_mirage.Make(S.TCP)
   module CA = Ca_certs_nss.Make(P)
 
+  let nameserver_of_string str =
+    let ( let* ) = Result.bind in
+    match String.split_on_char ':' str with
+    | "tls" :: rest ->
+      let str = String.concat ":" rest in
+      ( match String.split_on_char '!' str with
+      | [ nameserver ] ->
+        let* ipaddr, port = Ipaddr.with_port_of_string ~default:853 nameserver in
```

Comment:
Does this support ipv6? And how does it work with ipv6's `:` separator and port separator?

Reply:
You should check my recent PR about that: https://github.com/mirage/ocaml-ipaddr/pull/108
IF you want to specify a port with an IPv6 address, you must _expand_ the IPv6 address. For instance,
- `::1:8080` returns an IPv6 with the default port
- `0:0:0:0:0:0:0:1:8080` returns `::1` with the port `8080`